internal/model/plan.go

```diff
@@ -23,20 +24,70 @@ type Plan struct {
 
 	// The default quota values associated with the plan
 	PlanQuotaDefaults []PlanQuotaDefault `json:"plan_quota_defaults,omitempty"`
+
+	// The rates associated with the plan.
+	PlanRates []PlanRate `json:"plan_rates,omitempty"`
+}
+
+// Returns the currently active rate for a subscription plan. The active plan rate is the plan with the most recent
+// effective timestamp that occurs at or befor the curren time. This function assumes that the plan quota defaults are
+// sorted in ascending order by effective date.
+func (p *Plan) GetActivePlanRate() (*PlanRate, error) {
+	currentTime := time.Now()
+
+	// Find the active plan rate.
+	var activePlanRate *PlanRate
+	for _, pr := range p.PlanRates {
+		if pr.EffectiveDate.After(currentTime) {
+			break
+		}
+		activePlanRate = &pr
+	}
+
+	// It's an error for a plan not to have an active rate.
+	if activePlanRate == nil {
+		return nil, fmt.Errorf("no active rate found for subscription plan %s", *p.ID)
+	}
+
+	return activePlanRate, nil
 }
 
-// GetDefaultQuotaValue returns the default quota value associated with the resource type with the given name.
+// GetDefaultQuotaValue returns the default quota value associated with the resource type with the given name. This
+// funciton assumes that the plan quota defaults ar sorted in ascending order by effective date.
 func (p *Plan) GetDefaultQuotaValue(resourcetypeName string) float64 {
```

Comment:
I'm not sure if this is helpful or not, but it'd at least cut down lines of code if this function was implemented using the newly-added `GetDefaultQuotaValues` instead. Really the only thing this function is doing differently is a conditional on the resource type name rather than using that name as a key, and I believe it'll look at the same collection of quota defaults since they both just range over it and break when they pass `time.Now()`.

(the comment also has a couple typos, which doesn't really matter at all)

Reply:
Definitely helpful. Thanks for the suggestion!